### PR TITLE
support sonic-mgmt multi-dut gen-mg for LC

### PR DESCRIPTION
**Pull request by saravanansv**
1. support adding inband data in LC's minigraph like below
 VoqInbandInterfaces fields come from testbed.yaml (fields voq_inband_intf, voq_inband_type, voq_inband_ip) to lab/veos inventory files to minigraph.

2. System ports
Data flows from
a. port_config.ini (new fileds required are: numVoq, coreId, corePortId), (existing fields: name, speed)
b. switchId = running asic_id count across all linecards in the chassis
c. systemPortId = running systemport count across all linecards in the chassis
Each dut adds its own system-ports to its own ansible_facts.
config_sonic_basedon_testbed.yml loops through all duts system-ports to create all_sysports

3. DeviceProperty
   <a:DeviceProperty>
     <a:Name>SwitchType</a:Name>
     <a:Reference i:nil="true"/>
     <a:Value>voq</a:Value>
   </a:DeviceProperty>
   <a:DeviceProperty>
     <a:Name>MaxCores</a:Name>
     <a:Reference i:nil="true"/>
     <a:Value>16</a:Value>
   </a:DeviceProperty>
   <a:DeviceProperty>
     <a:Name>SwitchId</a:Name>
     <a:Reference i:nil="true"/>
     <a:Value>0</a:Value>
   </a:DeviceProperty>

a. switch type, maxcores are directly from testbed.yaml to inventory files to minigraph
b. start_switchid is calculated for each linecard and set into inventory files from TestbedProcessing.py based on num_asics from previous linecards

### Description of PR

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### CODE REVIEW

File: ansible/library/port_alias.py [hunk 1]
```diff
@@ -48,7 +48,8 @@
 ### Here are the expectation of files of device port_config.ini located, in case changed please modify it here
 FILE_PATH = '/usr/share/sonic/device'
 PORTMAP_FILE = 'port_config.ini'
-ALLOWED_HEADER = ['name', 'lanes', 'alias', 'index', 'asic_port_name', 'role', 'speed']
+ALLOWED_HEADER = ['name', 'lanes', 'alias', 'index', 'asic_port_name', 'role', 'speed',
+                  'coreid', 'coreportid', 'numvoq']
```
Comment (sanmalho-git) [inline ansible/library/port_alias.py:52]: Adding extra fields 'coreid', 'coreportid', and 'numvoq' in port_config.ini will require updating port_config.ini for all platforms across all vendors. Has this been discussed and approved in sonic chassis subgroup.  

Would sonic-buildimage be compatible with these added fields in port_config.ini.

File: ansible/TestbedProcessing.py [hunk 2]
```diff
@@ -416,6 +418,89 @@ def makeLab(data, devices, testbed, outfile):
...
+                            cardtype = dev.get("cardtype")
+                            if cardtype is not None:
+                               entry += "\tcardtype=" + cardtype
+                        except:
+                           print("\t\t" + host + ": cardtype not found")
+
+                        try: #get num_asics
+                            num_asic = dev.get("num_asic")
+                            if num_asic is not None:
+                               entry += "\tnum_asic=" + str( num_asic )
```
Comment (sanmalho-git) [inline ansible/TestbedProcessing.py:438]: I believe the variable name in the inventory/lab file should be 'num_asics', and not 'num_asic'

File: ansible/TestbedProcessing.py [hunk 3]
```diff
@@ -416,6 +418,89 @@ def makeLab(data, devices, testbed, outfile):
...
+                            if asics_host_ipv6 is not None:
+                               entry += "\tasics_host_ipv6=" + str( asics_host_ipv6 )
+                        except:
+                            print("\t\t" + host + " asics_host_ipv6 not found")
+
+                        try: #get voq_inband_ip
+                            voq_inband_ip = dev.get("voq_inband_ip")
+                            if voq_inband_ip is not None:
+                               entry += "\tvoq_inband_ip=" + str( voq_inband_ip )
+                        except:
```
Comment (sanmalho-git) [inline ansible/TestbedProcessing.py:474]: Need to also support voq_inband_ipv6 interface

File: ansible/library/port_alias.py (hunk 1 above)
Comment (saravanansv) [reply ansible/library/port_alias.py:52]: When I emailed sonic chassis subgroup there was no opposition. 

After looking at asic_port_name usage only in multi-asic-vs port_config.ini file,
(e.g device/virtual/x86_64-kvm_x86_64-r0/msft_multi_asic_vs/2/port_config.ini) I wanted to follow similar logic of adding coreid, coreportid, numvoq only to the applicable VoQ sku's port_config.ini and it shouldn't be required in all skus.

File: ansible/TestbedProcessing.py (hunk 2 above)
Comment (saravanansv) [reply ansible/TestbedProcessing.py:438]: the variable num_asics is already used in config_sonic_basedon_testbed.yml.
I can rename the new num_asic to num_asics if there is no conflict in my diffs with existing assumptions of num_asics

File: ansible/TestbedProcessing.py (hunk 3 above)
Comment (saravanansv) [reply ansible/TestbedProcessing.py:474]: Sure

File: ansible/library/port_alias.py (hunk 1 above)
Comment (sanmalho-git) [reply ansible/library/port_alias.py:52]: As long as compatibility is handled in sonic-buildimage - it is good.

File: ansible/TestbedProcessing.py [hunk 4]
```diff
@@ -416,6 +418,89 @@ def makeLab(data, devices, testbed, outfile):
...
+                            hwsku = dev.get("hwsku")
+                            if hwsku is not None:
+                               entry += "\thwsku=" + hwsku
+                        except:
+                            print("\t\t" + host + ": hwsku not found")
+
+                        try: #get cardtype
+                            cardtype = dev.get("cardtype")
+                            if cardtype is not None:
+                               entry += "\tcardtype=" + cardtype
```
Comment (sanmalho-git) [inline ansible/TestbedProcessing.py:431]: We use 'card_type' as the host variable to distinguish between linecard and supervisor card. Can we change 'cardtype' to 'card_type'.

File: ansible/TestbedProcessing.py [hunk 5]
```diff
@@ -457,6 +542,10 @@ def makeVeos(data, veos, devices, outfile):
                     try:
                         ansible_host = devices.get(host.lower()).get("ansible").get("ansible_host")
                         entry += "\tansible_host=" + ansible_host.split("/")[0]
+                        if dev.get("device_type") == "DevSonic":
+                            entry += "\ttype=" + dev.get("type")
+                            entry += "\thwsku=" + dev.get("hwsku")
+                            entry += "\tcardtype=" + dev.get("cardtype")
```
Comment (sanmalho-git) [inline ansible/TestbedProcessing.py:548]: We use 'card_type' as the host variable to distinguish between linecard and supervisor card. Can we change 'cardtype' to 'card_type'.

File: ansible/config_sonic_basedon_testbed.yml [hunk 6]
```diff
@@ -75,14 +75,29 @@
     delegate_to: localhost
 
   - name: find interface name mapping and individual interface speed if defined from dut
-    port_alias: hwsku="{{ hwsku }}"
+    port_alias:
+      hwsku: "{{ hwsku }}"
+      cardtype: "{{ cardtype | default('fixed') }}"
```
Comment (sanmalho-git) [inline ansible/config_sonic_basedon_testbed.yml:80]: We use 'card_type' as the host variable to distinguish between linecard and supervisor card. Can we change 'cardtype' to 'card_type'.

**Comment by anshuv-mfst** [comment]
Hi @yxieca - could you please assign MSFT reviewers, thanks.

**Comment by yxieca** [comment]
@arlakshm, @abdosi please review. Thanks!

File: ansible/TestbedProcessing.py (hunk 2 above)
Comment (saravanansv) [reply ansible/TestbedProcessing.py]: @SuvarnaMeenakshi should there be num_asic and also num_asics ? 
In most of the places it uses {{ num_asic }} and in one place it uses {{ num_asics }} .
Why do we need both ?   should it be just one ? 

ansible/config_sonic_basedon_testbed.yml:    port_alias: hwsku="{{ hwsku }}" num_asic="{{ num_asics }}"
ansible/roles/vm_set/tasks/kickstart_vm.yml:      num_asic: "{{ hostvars[vm_name]['num_asics'] | default(1) }}"
ansible/roles/vm_set/tasks/kickstart_vm.yml:              num_asic={{ num_asic }}
ansible/roles/vm_set/tasks/kickstart_vm.yml:              num_asic={{ num_asic }}
ansible/roles/vm_set/tasks/start_sonic_vm.yml:    num_asic: "{{ hostvars[dut_name]['num_asics'] | default(1) }}"
ansible/roles/vm_set/tasks/start_sonic_vm.yml:  port_alias: hwsku={{ hostvars[dut_name].hwsku }} num_asic={{ num_asic }}
ansible/roles/vm_set/tasks/start_sonic_vm.yml:             num_asic={{ num_asic }}
ansible/library/port_alias.py:      port_alias: hwsku='ACS-MSN2700' num_asic=1
ansible/library/port_alias.py:            num_asic=dict(type='int', required=False)

Comment (saravanansv) [reply ansible/TestbedProcessing.py]: ansible/roles/vm_set/tasks/kickstart_vm.yml uses {{ num_asic }} and not {{ num_asics }}
where is the num_asic coming from ?

Comment (SuvarnaMeenakshi) [reply ansible/TestbedProcessing.py]: > ansible/roles/vm_set/tasks/kickstart_vm.yml uses {{ num_asic }} and not {{ num_asics }}
> where is the num_asic coming from ?

https://github.com/Azure/sonic-mgmt/blob/master/ansible/roles/vm_set/tasks/kickstart_vm.yml#L53 - Here a variable "num_asic" is created which gets the data from "num_asics" in inventory file.
```
num_asic: "{{ hostvars[vm_name]['num_asics'] | default(1) }}"
```
The variable num_asic is referenced every where else in the file.
Ex: https://github.com/Azure/sonic-mgmt/blob/master/ansible/roles/vm_set/tasks/kickstart_vm.yml#L63 - here num_asic variable defined in line 53 above is used.

Comment (saravanansv) [reply ansible/TestbedProcessing.py]: num_asics and num_asic are very close and leads to confusion. 
should num_asic be renamed to num_asic_vm in kickstart_vm.yml and num_asic_dut in start_sonic_vm.yml ? or just use num_asics everywhere ?

**Comment by saravanansv** [comment]
Closed this PR. 
Will use https://github.com/Azure/sonic-mgmt/pull/3245 to have a common PR for both Fabric and LC gen-minigraph support for VoQ

**Comment by shubav** [comment]
> Closed this PR.
> Will use #3245 to have a common PR for both Fabric and LC gen-minigraph support for VoQ

Does this mean that 3245 needs a relook for the fabric side too? Or is just a matter or merging these changes into 3245? Just for my understanding. Thanks.

**Comment by saravanansv** [comment]
> > Closed this PR.
> > Will use #3245 to have a common PR for both Fabric and LC gen-minigraph support for VoQ
> 
> Does this mean that 3245 needs a relook for the fabric side too? Or is just a matter or merging these changes into 3245? Just for my understanding. Thanks.

I'm waiting for an all-pending tests to finish in 3245. 
After that will patch the diffs for Linecard from here to 3245. 
The existing diffs in 3245 for fabric will not be affected and continue to work.

File: ansible/TestbedProcessing.py (hunk 2 above)
Comment (SuvarnaMeenakshi) [reply ansible/TestbedProcessing.py]: > num_asics and num_asic are very close and leads to confusion.
> should num_asic be renamed to num_asic_vm in kickstart_vm.yml and num_asic_dut in start_sonic_vm.yml ? or just use num_asics everywhere ?

could use num_asics everywhere if that makes it easier, as the variable is locally used.

Comment (saravanansv) [reply ansible/TestbedProcessing.py]: I can rename all the num_asic to num_asics part of https://github.com/Azure/sonic-mgmt/pull/3245 or should that be done in a separate PR ?

Comment (SuvarnaMeenakshi) [reply ansible/TestbedProcessing.py]: I suggest having separate PR.

**Pull request closed without merging**